src/gt4sd/algorithms/generation/torchdrug/implementation.py

```diff
@@ -0,0 +1,192 @@
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+Task = Union[GCPNGeneration, AutoregressiveGeneration]
+
+"""
+Necessary because torchdrug silently overwrites the default nn.Module. This is quite
+invasive and causes significant side-effects in the rest of the code.
+See: https://github.com/DeepGraphLearning/torchdrug/issues/77
+"""
+nn.Module = nn._Module  # type: ignore
+
+
+class DummyDataset:
+    """A helper class to imitate a torchdrug dataset."""
+
+    def __init__(self, atom_types: List[int]):
+        self.atom_types = atom_types
+        self.transform = None
+
+
+class Generator:
+    """Implementation of a TorchDrug generator."""
+
+    solver: Engine
+    task: Task
+    num_sample: int = 32
+    max_resample: int = 16
+
+    def __init__(
+        self,
+        resources_path: str,
+        atom_types: List[int],
+        hidden_dims: List[int],
+        input_dim: int,
+        num_relation: int,
+        batch_norm: bool,
+        device: Optional[Union[torch.device, str]] = None,
+    ):
+        """A TorchDrug generation algorithm.
+
+        Args:
+            resources_path: path to the cache.
+            atom_types: list of atom types.
+            hidden_dims: list of hidden dimensions, one per layer.
+            num_relation: number of relations for the graph.
+            batch_norm: whether to use batch normalization.
+            device: device where the inference
+                is running either as a dedicated class or a string. If not provided is inferred.
+        """
+        self.device = device_claim(device)
+        self.resources_path = resources_path
+
+        self.model = models.RGCN(
+            input_dim=input_dim,
+            num_relation=num_relation,
+            hidden_dims=hidden_dims,
+            batch_norm=batch_norm,
+        )
+
+        self.dataset = DummyDataset(atom_types)
+
+    def load_model(self, resources_path: str):
+        """Load a pretrained TorchDrug model."""
+        self.solver.load(Path(resources_path).joinpath("weights.pkl"))
+
+    def sample(self) -> List[str]:
+        """Sample a molecule.
+
+        Returns:
+            a generated SMILES string wrapped into a list.
+        """
+        results = self.task.generate(
+            num_sample=self.num_sample, max_resample=self.max_resample
+        )
+        return results.to_smiles()
+
+
+class GCPNGenerator(Generator):
+    """
+    Interface for the GCPN model as implemented in TorchDrug.
+
+    For details see:
+    You, J. et al. (2018). Graph convolutional policy network for goal-directed
+    molecular graph generation. Advances in neural information processing systems, 31.
+
+    """
+
+    task: GCPNGeneration
+
+    input_dim: int = 18
+    num_relation: int = 3
+    batch_norm: bool = False
+    atom_types: List[int] = [6, 7, 8, 9, 15, 16, 17, 35, 53]
+    hidden_dims: List[int] = [256, 256, 256, 256]
+
+    def __init__(self, resources_path: str):
+        """
```

Comment:
add initial comment in docstring

Reply:
Done in [04fd6f6](https://github.com/GT4SD/gt4sd-core/pull/49/commits/04fd6f6d79c2703828c5c61c935673c2726eefad)